### PR TITLE
Azure RBAC feature toggle for AKS

### LINKED ISSUE #8535: Support for AAD based RBAC on Azure Kubernetes
### Community Note

* Please vote on this issue by adding a 👍 [reaction](https://blog.github.com/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/) to the original issue to help the community and maintainers prioritize this request
* Please do not leave "+1" or "me too" comments, they generate extra noise for issue followers and do not help prioritize the request
* If you are interested in working on this issue or have submitted a pull request, please leave a comment

### Description

This feature allows using Azure Active Directory based role assignments to control authorization checks on Azure Kubernetes Service clusters.

### New or Affected Resource(s)

* azurerm_kubernetes_cluster

### Potential Terraform Configuration

```hcl
resource "azurerm_kubernetes_cluster" "example" {
  name                = "example-aks1"
  location            = azurerm_resource_group.example.location
  resource_group_name = azurerm_resource_group.example.name
  dns_prefix          = "exampleaks1"
  
  aad_role_based_access_control = "true"

  default_node_pool {
    name       = "default"
    node_count = 1
    vm_size    = "Standard_D2_v2"
  }

  identity {
    type = "SystemAssigned"
  }

  tags = {
    Environment = "Production"
  }
}
```

### References
* https://docs.microsoft.com/en-us/azure/aks/manage-azure-rbac

### PR DESCRIPTION
**Pull request by aristosvo**
Fixes #8535

# Acceptance Tests

- [x] Existing RBAC tests succeed:

```bash
--- PASS: TestAccKubernetesCluster_roleBasedAccessControl (584.95s)
--- PASS: TestAccKubernetesCluster_roleBasedAccessControlAADManaged (613.16s)
--- PASS: TestAccKubernetesCluster_roleBasedAccessControlAAD (668.37s)
--- PASS: TestAccKubernetesCluster_roleBasedAccessControlAADUpdateToManaged (754.84s)
--- PASS: TestAccKubernetesCluster_roleBasedAccessControlAADManagedChange (881.14s)
PASS
```
- [x] Specific test is passing after enabling Preview Feature
```bash
# Test fails without registering for the Preview Feature
❯ TF_ACC=1 go test -v ./azurerm/internal/services/containers/ -run=TestAccKubernetesCluster_roleBasedAccessControlAzure -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccKubernetesCluster_roleBasedAccessControlAzure
=== PAUSE TestAccKubernetesCluster_roleBasedAccessControlAzure
=== CONT  TestAccKubernetesCluster_roleBasedAccessControlAzure
    testing.go:684: Step 0 error: errors during apply:
        
        Error: creating Managed Kubernetes Cluster "acctestaks210202221714650343" (Resource Group "acctestRG-aks-210202221714650343"): containerservice.ManagedClustersClient#CreateOrUpdate: Failure sending request: StatusCode=400 -- Original Error: Code="BadRequest" Message="aadProfile.enableAzureRBAC requires Microsoft.ContainerService/EnableAzureRBACPreview feature flag."
   
--- FAIL: TestAccKubernetesCluster_roleBasedAccessControlAzure (79.42s)
❯ az feature register --namespace "Microsoft.ContainerService" --name "EnableAzureRBACPreview"
❯ az feature list -o table --query "[?contains(name, 'Microsoft.ContainerService/EnableAzureRBACPreview')].{Name:name,State:properties.state}"

# Wait until "Registered"
❯ az provider register --namespace Microsoft.ContainerService
❯ TF_ACC=1 go test -v ./azurerm/internal/services/containers/ -run=TestAccKubernetesCluster_roleBasedAccessControlAzure -timeout 180m -ldflags="-X=github.com/terraform-providers/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccKubernetesCluster_roleBasedAccessControlAzure
=== PAUSE TestAccKubernetesCluster_roleBasedAccessControlAzure
=== CONT  TestAccKubernetesCluster_roleBasedAccessControlAzure
--- PASS: TestAccKubernetesCluster_roleBasedAccessControlAzure (615.32s)
PASS
ok      github.com/terraform-providers/terraform-provider-azurerm/azurerm/internal/services/containers      616.643s
```

### CODE REVIEW

**Comment by aristosvo** [comment]
@tombuildsstuff Is there a reason we don't want to merge this one?

File: azurerm/internal/services/containers/kubernetes_cluster_resource.go [hunk 1]
```diff
@@ -500,6 +500,14 @@ func resourceKubernetesCluster() *schema.Resource {
 										Optional: true,
 									},
 
+									"azure_rbac_enabled": {
```
Comment (katbyte) [inline azurerm/internal/services/containers/kubernetes_cluster_resource.go:503]: should this be
```suggestion
									"azure_ad_rbac_enabled": {
```

to make it obvious its AzureAD not some other form of azure RABC?

**Review by katbyte — CHANGES_REQUESTED** [review]
Thanks @aristosvo - aside from one comment this looks good

File: azurerm/internal/services/containers/kubernetes_cluster_resource.go (hunk 1 above)
Comment (aristosvo) [reply azurerm/internal/services/containers/kubernetes_cluster_resource.go:503]: I don't think so, as it actually is/enables Azure RBAC to be used for Kubernetes. What it means is that you can give a user a role in the portal (i.e. `Azure Kubernetes Service RBAC Viewer`) with scope (i.e. `$AKS_ID/namespaces/<namespace-name>`). 

The connection with Azure AD is that the cluster should have managed Azure AD integration enabled to facilitate this, but those mechanics are from a user point of view abstracted to Azure RBAC.

Comment (katbyte) [reply azurerm/internal/services/containers/kubernetes_cluster_resource.go:503]: 👍

**Review by katbyte — APPROVED** [review]
Thanks @aristosvo - LGTM 👍

**Comment by harripaalanen** [comment]
Azure RBAC is now GA (https://github.com/Azure/AKS/issues/1786). If I set azure_rbac_enabled=true is it able to update the existing AKS resource or is new resource creation forced?

**Comment by aristosvo** [comment]
> Azure RBAC is now GA (https://github.com/Azure/AKS/issues/1786). If I set azure_rbac_enabled=true is it able to update the existing AKS resource or is new resource creation forced?

New resource creation is forced currently, based on the docs this should be updated as after GA this is not necessary anymore

**Comment by aristosvo** [comment]
@harripaalanen FYI I checked whether I could loosen up that requirement, but it requires a Preview feature to be enabled:
```
updating Managed Kubernetes Cluster "acctestaks210526225032917802" (Resource Group "acctestRG-aks-210526225032917802"): containerservice.ManagedClustersClient#CreateOrUpdate: Failure sending request: StatusCode=0 
--
Original Error: Code="BadRequest" Message="aadProfile.enableAzureRBAC requires Microsoft.ContainerService/UpdateAzureRBACPreview feature flag."
```

For the time being I won't change it and keep an eye on it in the mean time.

**Comment by harripaalanen** [comment]
@aristosvo I think they fixed it. It should't require the preview feature anymore.

**Comment by aristosvo** [comment]
@harripaalanen Fixed in #12029

**Pull request merged**